### PR TITLE
kv/bulk: split at exiting key above ingested keys

### PR DESCRIPTION
**Pull request by dt**
If there is existing data above the span of an ingested SSTable, the
`AddSSTableResponse` now includes the key at which it starts. This can
be used by a caller to determine if and where it should split to avoid
its ingest span overlapping that existing data, since ingesting into an
empty span is generally ideal (we can scatter the empty RHS cheaply.)

This new field is then used by the sst batcher during ingest to split
off the existing data span if there is one above the file being added
any time we need to split to make room for a file. This ensures the span
into which that file is going is fully empty, and can be scattered.

Release note: none.

Jira issue: CRDB-14729

### CODE REVIEW

**Comment by dt** [comment]
master:
Screen Shot 2022-03-30 at 12 22 06 AM
Screen Shot 2022-03-30 at 12 22 28 AM

This change:
Screen Shot 2022-03-30 at 12 23 04 AM
Screen Shot 2022-03-30 at 12 22 45 AM

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go
Comment (erikgrinaker) [inline pkg/kv/kvserver/batcheval/cmd_add_sstable.go:277]: I think this requires taking out a read latch over this span. Is that going to be disruptive to other AddSSTable requests into the range?

File: pkg/roachpb/api.proto
Comment (erikgrinaker) [inline pkg/roachpb/api.proto:1770]: nit: to me, above sounds like across the span at higher timestamps. After? Following? RHS?

**Review by erikgrinaker — COMMENTED** [review]
Mostly reviewed the AddSSTable changes.

File: pkg/kv/kvserver/batcheval/cmd_add_sstable.go
Comment (dt) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go:277]: Ah, yeah, is an excellent point. I could get this key from AdminSplit instead and it already had a latch over the whole range. plus I only need this key when the range is full, which is when I'm going to split it, not every time I add to it. @nvanbenschoten I think this might suggest we revisit asking AdminSplit to reply with the first key in the new RHS (we're already looking for it for the stats choice too)

Comment (erikgrinaker) [reply pkg/kv/kvserver/batcheval/cmd_add_sstable.go:277]: That seems fine at a first glance, but I know @nvanbenschoten had some thoughts on this so I'll defer to him.

**Review by nvb — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 7 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @erikgrinaker, @nvanbenschoten, and @stevendanna)

___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-MzQITLHEbxRdBFVsg9x:-MzQITLHEbxRdBFVsg9y:bjxydgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc02efdb76c19b482319a9690d7f7908a6481593/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L271)):*
> ```Go
> 	reply.AvailableBytes = cArgs.EvalCtx.GetMaxBytes() - cArgs.EvalCtx.GetMVCCStats().Total() - stats.Total()
> 
> 	if args.CheckForDataAbove {
> ```

Could we add a comment explaining what this block is doing?
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mz1cC4ifSgnskQLxce3-r2-277:-MzQHVAx7cSPOtzofcUu:b-o6oj5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc02efdb76c19b482319a9690d7f7908a6481593/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L277)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

That seems fine at a first glance, but I know @nvanbenschoten had some thoughts on this so I'll defer to him. We could possibly just always include it in the result, since I don't think a seek here is a significant cost.
</blockquote></details>

Why does this require taking out read latches? This is a best-effort heuristic, just like the stats consultation is above. The scan is already inconsistent in that it's not consulting the next key's timestamp, checking whether it's an intent, or bumping the timestamp cache. So if we were going to grab latches, they would need to be non-MVCC read latches to force us to serialize with any write to the range. Do we need to do that, though? This wouldn't prevent someone from writing before this key after the AddSSTable returns.

I still prefer this being part of the AddSSTable API and not the AdminSplit API. If anything, we'd like to make AdminSplit less complex, not more. And not more in a way that prevents us from ever removing the complexity because it's tightly coupled to import performance. We've discussed viewing it as a "hint" in the past for unprivileged tenants.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @erikgrinaker, @nvanbenschoten, and @stevendanna)

___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mz1cC4ifSgnskQLxce3-r2-277:-MzQnZ-u4IQ_qtr3uxbY:bw3uj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc02efdb76c19b482319a9690d7f7908a6481593/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L277)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why does this require taking out read latches? This is a best-effort heuristic, just like the stats consultation is above. The scan is already inconsistent in that it's not consulting the next key's timestamp, checking whether it's an intent, or bumping the timestamp cache. So if we were going to grab latches, they would need to be non-MVCC read latches to force us to serialize with any write to the range. Do we need to do that, though? This wouldn't prevent someone from writing before this key after the AddSSTable returns.

I still prefer this being part of the AddSSTable API and not the AdminSplit API. If anything, we'd like to make AdminSplit less complex, not more. And not more in a way that prevents us from ever removing the complexity because it's tightly coupled to import performance. We've discussed viewing it as a "hint" in the past for unprivileged tenants.
</blockquote></details>

Ah, okay, if we don't need to latch that span then I think we can do it in addsst; it we did need to latch the whole range that'd be terrible, but if we don't, then it seems OK. It is still _slightly_ unfortunate that we need to fire up an iterator and seek to get the next key every time we add a file rather than just when we need that information, i.e. when it is actually full and time to split -- but I'm happy to do this for now if that's the preference.
 
Long term (i.e. once we're not looking at 22.1 back portable changes) I feel like what might be an improvement would be a new API here, with a single KV RPC like "AdminAdviseIncomingIngestion, which basically says "I'm going to add data to this span" which KV can then use to decide if it wants to do anything to prepare for that, e.g. to a) split a the beginning of that span b) optionally split off some non-empty upper edge of the range containing the span c) maybe scatter the span if it is empty or cheap, d) whatever else it wants to do (disable stats/retune the underlying lsm for that area/whatever). Then we could get individual splits and scatters out of the import code. Maybe the same RPC could optionally offer up a sample of keys and take over initial splits too.

File: pkg/roachpb/api.proto
Comment (dt) [reply pkg/roachpb/api.proto:1770]: ReturnFollowingNonEmptySpanStart ?

Comment (dt) [reply pkg/roachpb/api.proto:1770]: or maybe `ReturnFollowingLikelyNonEmptySpanStart` to emphasize the reduced promise we're making w.r.t. consistency and isolation

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @erikgrinaker, @nvanbenschoten, and @stevendanna)

___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-MzQITLHEbxRdBFVsg9x:-MzR7U0pDNpFaPZ-11JK:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc02efdb76c19b482319a9690d7f7908a6481593/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L271)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could we add a comment explaining what this block is doing?
</blockquote></details>

Added.

**Review by erikgrinaker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @nvanbenschoten, and @stevendanna)

___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mz1cC4ifSgnskQLxce3-r2-277:-MzVAxAA1kQe5sPGzZhj:bfqu0ov) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc02efdb76c19b482319a9690d7f7908a6481593/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L277)):*
> Why does this require taking out read latches?

I wasn't sure if latches were needed to preserve lower-level invariants, so erred on the side of caution. If they're only for external invariants then yeah, it's fine.
___
*[pkg/roachpb/api.proto, line 1770 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mz1cC4ifSgnskQLxce5-r2-1770:-MzVBC229T6n2B_0Ca2l:b-kgfdi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc02efdb76c19b482319a9690d7f7908a6481593/pkg/roachpb/api.proto#L1770)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Switched it to the verbose but hopefully descriptor ReturnFollowingLikelyNonEmptySpanStart
</blockquote></details>

That'll do, thanks.

**Comment by dt** [comment]
rebased on just the refactor from #79126 without its behavior change, so now this one isn't blocked and could merge first.

**Review by erikgrinaker — APPROVED** [review]
`AddSSTable` changes LGTM. But we should add a comment mentioning why it's safe to read outside the latch here.

**Comment by dt** [comment]
> add a comment mentioning why it's safe to read outside the latch here.

I thought the comment already laid that out?

**Review by nvb — COMMENTED** [review]
> Reviewed 4 of 4 files at r10, 1 of 1 files at r11, 1 of 1 files at r12, 3 of 3 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @nvanbenschoten, and @stevendanna)

___
*[pkg/kv/bulk/sst_batcher.go, line 428 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mzp9ENV7vXumfZvGmtZ:-Mzp9ENV7vXumfZvGmt_:bpdzuzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/91da90729be026801c14952496fbe6e05b68aac3/pkg/kv/bulk/sst_batcher.go#L428)):*
> ```Go
> 		// is full as well as because if we do not, that existing data will need to
> 		// be moved when we scatter the span we're splitting for ingestion.
> 		if len(b.lastRange.nextExistingKey) > 0 && end.Compare(b.lastRange.nextExistingKey) < 0 {
> ```

How will this behave in a mixed version cluster? A v21.2 node will return a nil NextExistingKey even if one exists and so we'll occasionally scatter a non-empty range. Maybe that's not the worst thing in the world and we're ok with the mixed-version performance being suboptimal.
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 302 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mzp9ycA9feGBu-15NxM:-Mzp9ycA9feGBu-15NxN:b-htmfsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/91da90729be026801c14952496fbe6e05b68aac3/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L302)):*
> ```Go
> 	// to latch the entire remainder of the range and/or look through intents in
> 	// addition, and instead just use this key-only iterator. If a caller actually
> 	// needs to know what data is there, it must issue its own a real Scan.
> ```

"its own a real Scan" something sounds off here.
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 303 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mzp97R43aCB01fKn4RH:-Mzp97R43aCB01fKn4RI:b-nd0pl) ([raw file](https://github.com/cockroachdb/cockroach/blob/91da90729be026801c14952496fbe6e05b68aac3/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L303)):*
> ```Go
> 	// addition, and instead just use this key-only iterator. If a caller actually
> 	// needs to know what data is there, it must issue its own a real Scan.
> 	if args.ReturnFollowingLikelyNonEmptySpanStart {
> ```

I must be missing something. Are we setting this flag on the client anywhere?
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 312 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mzp9o-78itevQbX2cul:-Mzp9o-78itevQbX2cum:bohi978) ([raw file](https://github.com/cockroachdb/cockroach/blob/91da90729be026801c14952496fbe6e05b68aac3/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L312)):*
> ```Go
> 		ok, err = existingIter.Valid()
> 		if err != nil {
> 			return result.Result{}, err
> ```

nit: consider wrapping this error with some context.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78523)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @erikgrinaker, @nvanbenschoten, and @stevendanna)

___
*[pkg/kv/bulk/sst_batcher.go, line 428 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mzp9ENV7vXumfZvGmtZ:-MzpBqYf0rf1rsFF-iQ0:b6pvgr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/91da90729be026801c14952496fbe6e05b68aac3/pkg/kv/bulk/sst_batcher.go#L428)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

How will this behave in a mixed version cluster? A v21.2 node will return a nil NextExistingKey even if one exists and so we'll occasionally scatter a non-empty range. Maybe that's not the worst thing in the world and we're ok with the mixed-version performance being suboptimal.
</blockquote></details>

yeah, we already scatter non-empty ranges in 21.2, since we just blindly scatter the RHS every 48mb added, so I think this is no worse. Indeed, in mixed version, I think we'll already never be sending splits or scatters because the addsstable replies won't include space remaining either. 
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mz1cC4ifSgnskQLxce3-r2-277:-MzR8cKp8-q8TaQeSQ4n:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc02efdb76c19b482319a9690d7f7908a6481593/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L277)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

> Why does this require taking out read latches?

I wasn't sure if latches were needed to preserve lower-level invariants, so erred on the side of caution. If they're only for external invariants then yeah, it's fine.
</blockquote></details>

:thumbsup: 
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 303 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mzp97R43aCB01fKn4RH:-MzpD54e2NJiawScag0N:bl1va4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/91da90729be026801c14952496fbe6e05b68aac3/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L303)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I must be missing something. Are we setting this flag on the client anywhere?
</blockquote></details>

whoops, lost in the most recent rebase to make our own addsstable requests vs DB method. nice catch.
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 312 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mzp9o-78itevQbX2cul:-MzpDEwY6NI1w-0Q9f3O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91da90729be026801c14952496fbe6e05b68aac3/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L312)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: consider wrapping this error with some context.
</blockquote></details>

Done.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-:-MzpF3SeAiGfJuFEkZCj:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @erikgrinaker, @nvanbenschoten, and @stevendanna)

___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 303 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mzp97R43aCB01fKn4RH:-MzpF46r9f0YIleXqmq9:b-vdwbdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/91da90729be026801c14952496fbe6e05b68aac3/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L303)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

whoops, lost in the most recent rebase to make our own addsstable requests vs DB method. nice catch.
</blockquote></details>

I'm sure you were planning to, but it's probably worth running a final IMPORT experiment once everything has landed as a sanity check that all of the different improvements have fallen into place as expected. That might also allow you to compare against v21.2 and make an assertion about the overall speedup between the releases.

**Comment by dt** [comment]
TFTRs!

bors r+

**Comment by dt** [comment]
blathers backport 22.1

**Review by dt — COMMENTED** [review]
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-:-MzpN-YdC9P7wbfN-JTo:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable.go, line 303 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/78523#-Mzp97R43aCB01fKn4RH:-MzpMmdG3J-GiolkQIVV:b-25e0sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/91da90729be026801c14952496fbe6e05b68aac3/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L303)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm sure you were planning to, but it's probably worth running a final IMPORT experiment once everything has landed as a sanity check that all of the different improvements have fallen into place as expected. That might also allow you to compare against v21.2 and make an assertion about the overall speedup between the releases.
</blockquote></details>

yep, was indeed and did an initial run this morning: tpce 50k trade imports on 10xn2-standard-8 clusters, 3x master, 3x21.2.7. 
master's avg time is 5947s, range is +/- 235s. 
21.2's avg time is 8720s, +/- 490s, 
so avg is ~30% faster. 

logs show 4-6min in split and scatters on master, and 35-50min on 21.2, so that explains almost all the perf difference. 21.2 doesn't break down split-time vs scatter time, so I don't quite the exact mix of RHS stats computation vs avoiding expensive scatters of non-empty RHS; IIRC the earlier benchmarks of master before the fixes showed it was roughly 1/3 RHS stats and 2/3 stupid scatters, but in in extreme cases, where we got stuck scattering a nearly full 500mb RHS over and over, that could be much more skewed (in some cases scatter was >85% of total import time).

Sorting is the next big use of time, and then there's some unaccounted time in flushing I need to track down, but this is a decent improvement for 22.1.0.

**Pull request merged**
